src/zkevm_specs/evm/execution/calldataload.py

```diff
@@ -0,0 +1,56 @@
+from ..instruction import Instruction, Transition
+from ..opcode import Opcode
+from ..table import RW, CallContextFieldTag, TxContextFieldTag
+from ..util import BufferReaderGadget
+from ...util.param import N_BYTES_WORD
+
+
+def calldataload(instruction: Instruction):
+    opcode = instruction.opcode_lookup(True)
+    instruction.constrain_equal(opcode, Opcode.CALLDATALOAD)
+
+    # offset is the 64-bit offset to start reading 32-bytes from start of calldata.
+    offset = instruction.rlc_to_fq_exact(instruction.stack_pop(), n_bytes=8)
+
+    tx_id = instruction.call_context_lookup(CallContextFieldTag.TxId, RW.Read)
+
+    if instruction.curr.is_root:
+        calldata_length = instruction.tx_context_lookup(tx_id, TxContextFieldTag.CallDataLength)
+        calldata_offset = 0
+    else:
+        calldata_length = instruction.call_context_lookup(CallContextFieldTag.CallDataLength)
+        calldata_offset = instruction.call_context_lookup(CallContextFieldTag.CallDataOffset)
+
+    src_addr = offset + calldata_offset
+    src_addr_end = calldata_length + calldata_offset
+
+    buffer_reader = BufferReaderGadget(
+        instruction, N_BYTES_WORD, src_addr, src_addr_end, N_BYTES_WORD
+    )
+
+    calldata_word = []
+    for idx in range(N_BYTES_WORD):
+        if buffer_reader.read_flag(idx):
+            if instruction.curr.is_root:
+                tx_byte = instruction.tx_calldata_lookup(tx_id, offset + idx)
+                buffer_reader.constrain_byte(idx, tx_byte)
+                calldata_word.append(int(tx_byte))
+            else:
+                mem_byte = instruction.memory_lookup(RW.Read, offset + idx)
+                buffer_reader.constrain_byte(idx, mem_byte)
+                calldata_word.append(int(mem_byte))
+        else:
+            buffer_reader.constrain_byte(idx, 0)
+            calldata_word.append(0)
```

Comment:
I'm not sure you can append here directly. The padding is done with 0's on the left (If we consider BigEndian). As you se here: https://github.com/ethereum/go-ethereum/blob/8fddf27a989e246659fd018ea9be37b2b4f55326/core/vm/common.go#L54 (This is with LE and bytes).

Reply:
My understanding is that if the buffer reader has marked that index as "no data available", that needs to be padded with `0` (in the case of `CALLDATALOAD`), hence:
```python
if buffer_reader.read_flag(idx):
    // append from tx calldata (root) or memory (internal)
else:
    // append 0 (this is the right padding)
```

Reply:
@CPerezz Hmm, I am a little confused because I had followed the playground [setup here](https://www.evm.codes/playground?unit=Wei&callData=0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF&codeType=Mnemonic&code=%27~1w0yzz~2w31y%27~%2F%2F%20Example%20z%5CnyzCALLDATALOADwzPUSH1%20%01wyz~_).

Could you also have a look at it?

Reply:
kk it's my bad again. The values are padded correctly as for LE you padd to the right obviously..

Thanks for the link @roynalnaruto and sorry for the noise! :smile: